src/Features/LanguageServer/Protocol/Handler/Diagnostics/DocumentPullDiagnosticHandler.cs

```diff
@@ -69,13 +71,24 @@ protected override bool TryCreateUnchangedReport(TextDocumentIdentifier identifi
         protected override DiagnosticTag[] ConvertTags(DiagnosticData diagnosticData, bool isLiveSource)
             => ConvertTags(diagnosticData, isLiveSource, potentialDuplicate: false);
 
+        protected override VSInternalDiagnosticReport[]? CreateReturn(BufferedProgress<VSInternalDiagnosticReport[]> progress)
+            => progress.GetFlattenedValues();
+
         protected override ValueTask<ImmutableArray<IDiagnosticSource>> GetOrderedDiagnosticSourcesAsync(
             VSInternalDocumentDiagnosticsParams diagnosticsParams, RequestContext context, CancellationToken cancellationToken)
+            => new(GetDiagnosticSource(diagnosticsParams, context) is { } diagnosticSource ? [diagnosticSource] : []);
+
+        private IDiagnosticSource? GetDiagnosticSource(VSInternalDocumentDiagnosticsParams diagnosticsParams, RequestContext context)
         {
             var category = diagnosticsParams.QueryingDiagnosticKind?.Value;
 
+            // TODO: Implement as extensibility point. https://github.com/dotnet/roslyn/issues/72896
+
             if (category == PullDiagnosticCategories.Task)
-                return new(GetDiagnosticSources(diagnosticKind: default, nonLocalDocumentDiagnostics: false, taskList: true, context, GlobalOptions));
+                return context.GetTrackedDocument<Document>() is { } document ? new TaskListDiagnosticSource(document, GlobalOptions) : null;
+
+            if (category == PullDiagnosticCategories.EditAndContinue)
```

Comment:
currently these won't ever show in VSCode (because we only register the kind for VS diagnostics).

For VSCode, buckets are named 'identifiers', and they can only be registered dynamically like we do here - https://github.com/dotnet/roslyn/blob/b67df4c5e39ac1a104b2e573542211d4d6b1cd1d/src/Features/LanguageServer/Protocol/Handler/Diagnostics/Public/PublicDocumentPullDiagnosticsHandler_IOnInitialized.cs#L43

And unfortunately it has a different field name.  If you add add the identifier, in the Public doc pull handler, you'll also need to add the code to check for the identifier and create a source for it.